File: ui_test/local_env.py
from __future__ import annotations

from pathlib import Path
from typing import Any

from ui_test.env_loader import _parse_env_file, require_keys
from ui_test.project_profile import load_cheatsheet
from ui_test.project_paths import agent_dir, ENV_EXAMPLE, ENV_FILE, ENV_LOCAL


def cheatsheet_env_files(project: Path) -> tuple[str, ...]:
    sheet = load_cheatsheet(project)
    local = sheet.get("local") if isinstance(sheet.get("local"), dict) else {}
    files: list[str] = []
    raw = local.get("env_files")
    if isinstance(raw, list):
        files = [str(f).strip() for f in raw if str(f).strip()]
    env_file = str(local.get("env_file") or "").strip()
    if env_file:
        files.insert(0, env_file)
    if not files:
        files = ["micro-services/admin/.env", f"{agent_dir(project).name}/.env"]
    return tuple(dict.fromkeys(files))


def cheatsheet_required_env(project: Path) -> tuple[str, ...]:
    sheet = load_cheatsheet(project)
    local = sheet.get("local") if isinstance(sheet.get("local"), dict) else {}
    raw = local.get("required_env")
    if isinstance(raw, list):
        keys = [str(k).strip() for k in raw if str(k).strip()]
        if keys:
            return tuple(dict.fromkeys(keys))
    return ("DATABASE_URL",)


def load_merged_local_env(project: Path) -> dict[str, str]:
    merged: dict[str, str] = {}
    for rel in cheatsheet_env_files(project):
        path = project / rel if not Path(rel).is_absolute() else Path(rel)
        merged.update(_parse_env_file(path))
    return merged


def _env_example_path(env_path: str) -> str:
    if env_path.endswith(".env"):
        return f"{env_path[:-4]}.env.example"
    return f"{env_path}.example"


def _primary_env_file(env_files: tuple[str, ...]) -> str:
    normalized = [f.replace("\\", "/") for f in env_files]
    app_files = [f for f in normalized if ".agent/" not in f]
    if app_files:
        return app_files[-1]
    return normalized[-1] if normalized else f"{AGENT_DIR}/{ENV_FILE}"


def local_env_status(project: Path) -> dict[str, Any]:
    base = agent_dir(project)
    env_files = cheatsheet_env_files(project)
    required = cheatsheet_required_env(project)
    merged = load_merged_local_env(project)
    missing = require_keys(merged, list(required))
    primary_env = _primary_env_file(env_files)
    primary_example = _env_example_path(primary_env)

    file_status: list[dict[str, Any]] = []
    for rel in env_files:
        path = project / rel if not Path(rel).is_absolute() else Path(rel)
        file_status.append({"path": rel, "exists": path.is_file()})

    primary_path = project / primary_env if not Path(primary_env).is_absolute() else Path(primary_env)
    example_path = project / primary_example if not Path(primary_example).is_absolute() else Path(primary_example)
    return {
        "ready": len(missing) == 0,
        "missing": missing,
        "required": list(required),
        "env_files": file_status,
        "has_env": primary_path.is_file(),
        "has_env_local": (base / ENV_LOCAL).is_file(),
        "has_env_example": example_path.is_file(),
        "env_path": primary_env,
        "env_example_path": primary_example,
        "env_local_path": primary_env,
    }


def format_local_env_hint(project: Path, missing: list[str]) -> str:
    status = local_env_status(project)
    keys = ", ".join(missing)
    if status["has_env_example"]:
        return (
            f"Missing {keys}. Copy {status['env_example_path']} -> {status['env_path']} "
            f"and set values (see cheatsheet notes)."
        )
    return f"Missing {keys} in {status['env_path']} or cheatsheet env_files."

File: test-runner/server/local-env.ts
import fs from "node:fs";
import path from "node:path";

function parseEnvFile(content: string): Record<string, string> {
  const result: Record<string, string> = {};
  for (const line of content.split("\n")) {
    const stripped = line.trim();
    if (!stripped || stripped.startsWith("#")) continue;
    const eq = stripped.indexOf("=");
    if (eq <= 0) continue;
    const key = stripped.slice(0, eq).trim();
    let value = stripped.slice(eq + 1).trim();
    if (value.length >= 2 && value[0] === value.at(-1) && (value[0] === '"' || value[0] === "'")) {
      value = value.slice(1, -1);
    }
    result[key] = value;
  }
  return result;
}

function parseCheatsheetEnvFiles(cheatsheet: string): string[] {
  const files: string[] = [];
  const block = cheatsheet.match(/env_files:\s*\n((?:\s+-\s+.+\n?)+)/);
  if (block) {
    for (const line of block[1].split("\n")) {
      const m = line.match(/^\s+-\s+(.+)$/);
      if (m) files.push(m[1].trim());
    }
  }
  const single = cheatsheet.match(/^\s*env_file:\s*(.+)$/m);
  if (single) files.unshift(single[1].trim());
  if (!files.length) {
    files.push(".agent/.env", ".agent/.env.local");
  }
  return [...new Set(files)];
}

function parseRequiredEnv(cheatsheet: string): string[] {
  const block = cheatsheet.match(/required_env:\s*\n((?:\s+-\s+.+\n?)+)/);
  if (block) {
    const keys: string[] = [];
    for (const line of block[1].split("\n")) {
      const m = line.match(/^\s+-\s+(.+)$/);
      if (m) keys.push(m[1].trim());
    }
    if (keys.length) return keys;
  }
  return ["DATABASE_URL"];
}

export function readLocalEnvStatus(projectPath: string) {
  const resolved = path.resolve(projectPath);
  const agentDir = path.join(resolved, ".agent");
  const cheatsheetPath = path.join(agentDir, "cheatsheet.yaml");
  const cheatsheet = fs.existsSync(cheatsheetPath) ? fs.readFileSync(cheatsheetPath, "utf8") : "";
  const envFiles = parseCheatsheetEnvFiles(cheatsheet);
  const required = parseRequiredEnv(cheatsheet);

  const merged: Record<string, string> = {};
  const fileStatus: { path: string; exists: boolean }[] = [];
  for (const rel of envFiles) {
    const filePath = path.isAbsolute(rel) ? rel : path.join(resolved, rel);
    fileStatus.push({ path: rel, exists: fs.existsSync(filePath) });
    if (fs.existsSync(filePath)) {
      Object.assign(merged, parseEnvFile(fs.readFileSync(filePath, "utf8")));
    }
  }

  const missing = required.filter((key) => !String(merged[key] ?? "").trim());
  const envExample = path.join(agentDir, ".env.example");
  const envLocal = path.join(agentDir, ".env.local");

  return {
    ready: missing.length === 0,
    missing,
    required,
    env_files: fileStatus,
    has_env_example: fs.existsSync(envExample),
    has_env_local: fs.existsSync(envLocal),
    env_example_path: ".agent/.env.example",
    env_local_path: ".agent/.env.local",
    local_base_url: cheatsheet.match(/^\s*base_url:\s*(.+)$/m)?.[1]?.trim() ?? "",
  };
}
